File: core/causal_intelligence/semantic_runtime_mutation_lineage_engine.py
from __future__ import annotations

from typing import Any, Dict, List


MAX_MUTATIONS = 1000


def build_mutation_lineage(
    runtime_ir: Dict[str, Any],
) -> Dict[str, Any]:
    keys = sorted(runtime_ir.keys())[:MAX_MUTATIONS]
    lineage = [
        {
            "step": idx,
            "key": key,
            "causal_action": "observe",
        }
        for idx, key in enumerate(keys)
    ]
    return {
        "mutation_lineage": lineage,
        "bounded": True,
    }

File: src/causal_intelligence/semanticRuntimeMutationLineageEngine.ts
/**
 * Converted from Python: core/causal_intelligence/semantic_runtime_mutation_lineage_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_MUTATIONS: any = 1000;
export function buildMutationLineage(runtime_ir: any): any {
  var keys: any = py.slice(py.sorted(py.keys(runtime_ir)), null, MAX_MUTATIONS);
  var lineage: any = py.enumerate(keys).map(([idx, key]: any) => ({"step": idx, "key": key, "causal_action": "observe"}));
  return {"mutation_lineage": lineage, "bounded": true};
}
